resetProfile() fix #3692

Edru2 opened the pull request:
#### Brief overview of PR changes/additions
Order of reenabling/compiling units (trigger/alias...) was wrong and caused problems for resetProfile()

#### Motivation for adding to Mudlet
fixes #3692 

#### Other info (issues closed, discussion etc)

Edru2 committed:
fixes resetProfile()

units need to be enabled to be compiled
the order was wrong

vadi2 reviewed (approved):
This looks good to me :+1:

Eraene commented:
This immediately causes all running timers to start erroring with nil values until saved, still. I am assuming this is not an intended side effect?

Edru2 commented:
Yeah, there is a comment in the code which says timers are not compiled.
Maybe someone knows what's the reason for that?

It's easily fixable though.
will commit

Edru2 committed:
timers were not re-compiled

SlySven reviewed (dismissed):
Um, the timers not being compiled is an effect of e0ba28d4729f97f69cd91b178886ad6e7438d9a9 which was done by Heiko in 2010-12-28. I do not know all the details but IIRC some aspects of the timers are managed centrally by the `mudlet` class aren't they - is it possible that they get compiled automagically in some manner that means they should not have to be recompiled explicitly here?

Yeah - looking through the code `(void) mudlet::slot_timer_fires()` identifies the `TTimer` instance involved when the base class `QTimer` from which it is derived fires and then that calls the `TTimer::execute()` which arranges for the timer code (if any) to be compiled if it is needed (non-empty Lua script) - the methodology is different depending on whether it is a temporary or a permanent one - but both seem to do the compilation.

Therefore your premise that `getTimerUnit()->compileAll();` is needed seems highly suspect to me.

demonnic commented:
Why then are the timers throwing nil errors until you go in and force them to recompile?

Edru2 commented:
As @demonnic  and @Eraene  already mentioned ```getTimerUnit()->compileAll();``` seems to be needed if there are active (perm)timers. If it's not present this happens (after using resetProfile()):
![Screenshot from 2020-08-13 09-05-27](https://user-images.githubusercontent.com/60551052/90105854-df2dee80-dd46-11ea-9d58-06e4beab10d2.png)

SlySven commented:
Something is getting broken (or `nil`ed) in the Lua sub-system - but I'm not convinced that `getTimerUnit()->compileAll()` is the only way to fix it and there may be side effects to doing that. (Like the unrelated - except that it also involved timers and them getting recompiled whenever they are merely *viewed* in the editor - #2510 )

vadi2 commented:
Well, let's test the latest links and see if there are any sideffects. If everything works fine as it should, then this is good to go :+1:

Edru2 commented:
> Something is getting broken (or `nil`ed) in the Lua sub-system - but I'm not convinced that `getTimerUnit()->compileAll()` is the only way to fix it and there may be side effects to doing that. (Like the unrelated - except that it also involved timers and them getting recompiled whenever they are merely _viewed_ in the editor - #2510 )

I agree with you on #2510 that clicking save shouldn't reset a timer.

I don't see it as an issue in this case here.
I mean the command is called resetProfile() and that at least in my case lets me expect that it resets the timers.

As far as I see this part of the code is only in use if someone uses the lua function resetProfile() so it shouldn't affect any other parts of Mudlet.

> Well, let's test the latest links and see if there are any sideffects. If everything works fine as it should, then this is good to go


:+1:

Edru2 committed:
TimerUnit compileAll will now really compile all timers

TimerUnit compileAll didn't compile all timers but only the ones in the rootnode what means that timers in a folder weren't compiled at all

vadi2 commented:
@Edru2 has it been working well for you?

Edru2 commented:
Just noticed that eventHandler (the Editor ones) don't get reloaded.
Not a big issue, because most of the people using resetProfile() probably use it for testing their GUI offline but I will still look into fixing it.

Didn't notice any other issue.

vadi2 commented:
OK. @SlySven do you want to do something about your timers concern?

Edru2 committed:
fix eventhandler not loading after resetProfile()

Edru2 commented:
fixed the eventhandler issue.
Didn't notice anything else when testing.
for me it is okay now.

vadi2 commented:
@Eraene @demonnic could you test the latest incarnation? Good to make resetprofile trustworthy again.

SlySven commented:
> OK. @ SlySven do you want to do something about your timers concern?

Nope, if this PR does what it says I don't think I need to worry too much. I thought it would be unnecessary to recompile the timers but I'm hard pressed to see what any side effects there might be. Can I just reserve the right to say "I *told* you there might be a side-effect of doing that...!" if it does go pear-shaped later? :grinning:

Delwing commented:
Awesome! :+1: Thanks for looking into this.
